### PR TITLE
[FIX] website_livechat: ensure session data integrity during visitor lifecycle

### PR DESCRIPTION
**Pull request by dbeguin**
This commit fixes two issues encountered during the visitor lifecycle.

1. When a visitor arrives on the website on a non tracked page, a
website.visitor is not created yet, but the visitor can still start a livechat
session. When navigating to a tracked page, a website.visitor is created but the
lifechat session already started is not recovered and the discussion is lost at
visitor's side. This commit links the already started livechat session to the
newly created visitor to ensure that the conversation can still continue
normally.

2. When a visitor logs in, his attributed visitor is linked to the partner. But
there can be only one visitor per partner. If there was already a visitor linked
to the partner, the later visitor's livechat session are updated with the
partner and the visitor is then deleted. But the visitor's livechat sessions are
not linked to the main partner visitor. So when looking at all the session the
partner had, we can only see the main visitor's livechat session.
This commit copy the livechat session of that later visitor to the main partner
visitor to ensure keeping the complete livechat history for each partner.

Task ID: 2460892

### CODE REVIEW

File: addons/website_livechat/models/website_visitor.py [hunk 1]
```diff
@@ -99,4 +100,22 @@ def _link_to_partner(self, partner, update_values=None):
                 (3, self.env.ref('base.public_partner').id),
                 (4, partner.id),
             ]
+            # copy sessions of the secondary visitors to the main partner visitor.
+            main_partner_visitor = self.env['website.visitor'].sudo().search(
```
Comment (tde-banana-odoo) [inline addons/website_livechat/models/website_visitor.py:104]: Shouldn't it be in ``_link_to_visitor`` whose purpose is to link visitors ?

File: addons/website_livechat/models/website_visitor.py [hunk 2]
```diff
@@ -99,4 +100,22 @@ def _link_to_partner(self, partner, update_values=None):
...
+                main_partner_visitor.mail_channel_ids |= self.mail_channel_ids
         super(WebsiteVisitor, self)._link_to_partner(partner, update_values=update_values)
+
+    def _create_visitor(self):
+        visitor = super(WebsiteVisitor, self)._create_visitor()
+        mail_channel_uuid = json.loads(request.httprequest.cookies.get('im_livechat_session', '{}')).get('uuid')
+        if mail_channel_uuid:
+            mail_channel = request.env["mail.channel"].search([("uuid", "=", mail_channel_uuid)])
+            visitor.mail_channel_ids = [(4, mail_channel.id)]
+            mail_channel.sudo().write({
```
Comment (tde-banana-odoo) [inline addons/website_livechat/models/website_visitor.py:118]: Why searching as normal and writing as sudo ?

File: addons/website_livechat/models/website_visitor.py [hunk 3]
```diff
@@ -99,4 +100,22 @@ def _link_to_partner(self, partner, update_values=None):
...
+            if main_partner_visitor:
+                main_partner_visitor.mail_channel_ids |= self.mail_channel_ids
         super(WebsiteVisitor, self)._link_to_partner(partner, update_values=update_values)
+
+    def _create_visitor(self):
+        visitor = super(WebsiteVisitor, self)._create_visitor()
+        mail_channel_uuid = json.loads(request.httprequest.cookies.get('im_livechat_session', '{}')).get('uuid')
+        if mail_channel_uuid:
+            mail_channel = request.env["mail.channel"].search([("uuid", "=", mail_channel_uuid)])
+            visitor.mail_channel_ids = [(4, mail_channel.id)]
```
Comment (tde-banana-odoo) [inline addons/website_livechat/models/website_visitor.py:117]: Couldn't we write on ``livechat_visitor_id`` field of channel instead ? As we alreado do a write herebelow.

**Review by tde-banana-odoo — COMMENTED** [review]
Review

**Comment by dbeguin** [comment]
@tde-banana-odoo : You vere all vright !

File: addons/website_livechat/models/website_visitor.py [hunk 4]
```diff
@@ -92,6 +93,17 @@ def action_send_chat_request(self):
                 notifications.append([(self._cr.dbname, 'res.partner', operator.partner_id.id), mail_channel_info])
             self.env['bus.bus'].sendmany(notifications)
 
+    def _link_to_visitor(self, target, keep_unique=True):
+        """ Copy sessions of the secondary visitors to the main partner visitor. """
+        if target.partner_id:
+            main_partner_visitor = self.env['website.visitor'].sudo().search(
```
Comment (tde-banana-odoo) [inline addons/website_livechat/models/website_visitor.py:99]: Actually main_partner_visitor should be target. If I remember this method, target is the one where we merge data, so normally no need to re-compute something here.

**Review by tde-banana-odoo — COMMENTED** [review]
Gargamblaps !

**Comment by tde-banana-odoo** [comment]
@robodoo r+

**Pull request closed without merging**